### PR TITLE
CI: Add blank app fixture runner and appium tests

### PR DESCRIPTION
**Pull request by descorp**
# Changes

## Description
  This PR introduces a new End-to-End (E2E) testing infrastructure designed to validate the library against fresh, "blank" React Native and Expo projects. It also optimizes several CI workflows for better performance and observability.

###  🧪 End-to-End Testing (Appium)

   - **New E2E Suite**: Added a smoke test suite in `e2e-tests/` that uses `Appium/WebdriverIO` to verify the library loads correctly in newly initialized projects.
   - **On-Demand Validation**: Introduced `blank_test.yml`, allowing manual or label-triggered testing of specific React Native versions or Expo SDKs.
   - **Fixture Generation**: Added scripts to autonomously scaffold test applications, inject credentials, and run native builds for both platforms.

###  🚀 CI/CD Enhancements

   - **Composite Actions**: Refactored Android emulator setup into a reusable local action (`.github/actions/setup-android-emulator`) to reduce duplication.
   - **Faster Sonar Scans**: Migrated `run_sonar.yml` to Ubuntu and switched to the npm-based Sonar scanner, significantly reducing execution time.
   - **Improved Debugging**: Added automatic uploads for `xcodebuild` logs upon test failures.
   - **Dynamic Simulator Resolution**: Replaced hardcoded iOS device names with a robust resolution script that finds the best available simulator runtime.

###  🧹 Code Quality & Cleanup
   - **Linting**: Simplified eslint.config.mjs by removing redundant Prettier rules now handled by .prettierrc.json.

### CODE REVIEW

File: e2e-tests/template/App.tsx.template [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+  const { start } = useAdyenCheckout();
+
+  return (
+    <View>
+      <Button
+        title="Open DropIn"
+        onPress={async () => {
+          start('dropin');
+        }}
+      />
```
[the remark this answers is not in the document]
Comment (descorp) [reply e2e-tests/template/App.tsx.template:100]: Addressed in [4c813a6](https://github.com/Adyen/adyen-react-native/pull/880/commits/4c813a62ac179b154f327cbeeffe842d9e1154b1)

File: scripts/check-app.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    logLevel: 'error',
+  });
+
+  try {
+    // 1. Define Selector
+    // Android: Locates by visible text
+    // iOS: Locates by Accessibility ID (which maps to 'title' prop in RN Buttons)
+    const btnSelector = isAndroid 
+      ? '//*[@text="Open DropIn"]' 
+      : '~Open DropIn';
```
[the remark this answers is not in the document]
Comment (descorp) [reply scripts/check-app.js]: Addressed in [4c813a6](https://github.com/Adyen/adyen-react-native/pull/880/commits/4c813a62ac179b154f327cbeeffe842d9e1154b1)

File: scripts/fixture_setup.sh [hunk 3]
```diff
@@ -0,0 +1,189 @@
+#!/bin/bash
+
+# Default values
+expo=''
+name="tested_app"
+
+# Parse flags
+while getopts r:g:t:e flag
```
[the remark this answers is not in the document]
Comment (descorp) [reply scripts/fixture_setup.sh]: Resolved

File: scripts/fixture_android.sh [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+EOF
+fi
+
+echo "== Build Android"
+cd android || exit
+./gradlew assembleDebug
+cd ..
+
+echo "== Run Appium Tests"
+node ../scripts/check-app.js --platform android
```
[the remark this answers is not in the document]
Comment (descorp) [reply scripts/fixture_android.sh]: Resolved

File: scripts/fixture_ios.sh [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  pod install --repo-update
+  cd ..
+  SCHEME="TestProject"
+fi
+
+echo "== Build iOS"
+xcodebuild -workspace "ios/$SCHEME.xcworkspace" -scheme "$SCHEME" -configuration Debug -sdk iphonesimulator -destination "platform=iOS Simulator,name=$device_name,OS=$os_version" -quiet
+
+echo "== Run Appium Tests"
+node ../scripts/check-app.js --platform ios
```
[the remark this answers is not in the document]
Comment (descorp) [reply scripts/fixture_ios.sh]: Resolved

File: scripts/fixture_setup.sh [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+do
+    case "${flag}" in
+        r) rn=${OPTARG}
+           name="${name}_RN${rn}";;
+        g) gradle=${OPTARG}
+           name="${name}_GR${rn}";;
+        t) target=${OPTARG}
+           name="${name}_TR${rn}";;
```
[the remark this answers is not in the document]
Comment (descorp) [reply scripts/fixture_setup.sh]: Resolved

File: scripts/inject_secrets.sh [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+if [ -z "$ADYEN_CLIENT_KEY" ] || [ -z "$ADYEN_PUBLIC_KEY" ]; then
+  echo "Error: ADYEN_CLIENT_KEY and ADYEN_PUBLIC_KEY environment variables are missing."
+  exit 1
+fi
+
+cd "$name" || exit
+
+echo "== Injecting Secrets"
+sed "s|__CLIENT_KEY__|$ADYEN_CLIENT_KEY|g" App.tsx > App.tsx.tmp && mv App.tsx.tmp App.tsx
+sed "s|__PUBLIC_KEY__|$ADYEN_PUBLIC_KEY|g" App.tsx > App.tsx.tmp && mv App.tsx.tmp App.tsx
```
[the remark this answers is not in the document]
Comment (descorp) [reply scripts/inject_secrets.sh]: Resolved

File: .github/workflows/blank_test.yml [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+      - name: Setup Appium
+        run: |
+          npm i -g appium
+          appium driver install xcuitest
+
+      - name: Start Appium
+        run: |
+          appium --port 4723 --log-level error &
+          for i in {1..90}; do
+            if curl -sf http://127.0.0.1:4723/status >/dev/null; then
```
Comment (atmamont) [inline .github/workflows/blank_test.yml]: is this waiting for a server to start up and be ready?

**Review by atmamont — COMMENTED** [review]
Amazing investment into reliability, Vlad.
I have some concerns regarding injecting secrets in the script runs using `sed` as it might be a sec vulnerability as they will be visible in logs.
I am not a fan of bash scripts as they are hard to maintain for anyone but author, let me take some time to think of what could be improved in terms of readability.

File: scripts/fixture_android.sh [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+for i in {1..30}; do
+  nc -z 127.0.0.1 8081 && break
+  sleep 1
+done
+
+echo "== Reverse Metro Port"
+adb reverse tcp:8081 tcp:8081 || true
+
+echo "== Launch App"
+adb shell am start -n "$APP_PACKAGE/$APP_ACTIVITY" >/dev/null 2>&1 || true
```
Comment (atmamont) [inline scripts/fixture_android.sh]: `|| true` means it will pass it the command fails, why wouldn't we fail early here?

File: scripts/fixture_ios.sh [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+xcrun simctl boot "$UDID" || true
+xcrun simctl bootstatus "$UDID" -b
+xcrun simctl install "$UDID" "build/Build/Products/Debug-iphonesimulator/$SCHEME.app"
+
+echo "== Run Appium Tests"
+export PLATFORM_NAME=ios
+export IOS_UDID="$UDID"
+export IOS_DEVICE_NAME="$device_name"
+export IOS_PLATFORM_VERSION="$os_version"
+node ./check-app.js
```
Comment (atmamont) [inline scripts/fixture_ios.sh]: This is quite a complex setup and is not scalable for other platforms(at least flutter). Let's think of alternative solutions? 
Using meta repository pipelines that run nightly or upon a new release could be more scalable, as it uses appium repo as a submodule and you can get rid of all wrapper and setup code that would be a hell to maintain.

File: scripts/fixture_android.sh [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+cd ..
+
+APP_PACKAGE=$(grep -m 1 -E 'applicationId\s+"' android/app/build.gradle | awk -F '"' '{print $2}')
+APP_ACTIVITY=${APP_ACTIVITY:-.MainActivity}
+if [ -z "${APP_PACKAGE:-}" ]; then
+  APP_PACKAGE=com.testproject
+fi
+
+echo "== Start Metro"
+yarn start --port 8081 >/dev/null 2>&1 &
```
Comment (atmamont) [inline scripts/fixture_android.sh]: This could leave bundler running if script fails (I am not sure but it can easily burn github runner paid minutes). It's a general recommendation to catch such things, here is one suggestion from LLM:
```
cleanup() {
  pkill -f "node.*metro" || true
}
trap cleanup EXIT
```

**Comment by atmamont** [comment]
@Robert-SD we need your input here

While this is definitely a nice improvement to react native, I feel it would be hard to scale to all platforms.

**Key issues:**
1. Duplicates existing infrastructure: Appium test suite in adyen-checkout-e2e-testing that covers iOS, Android, and Flutter with:
   - WebDriverIO + Page Object Model architecture (Robert is working on refactoring the engine to use Kotlin instead)
   - Centralized element mapping system
   - 7+ comprehensive test specs (not just smoke tests) and more to come when we scale this up in 2026
   - Existing CI/CD workflows that only need to be triggered with a provided app binary (it's MyStore now, could be any, as warmup phase is abstracted and you can write your own set of steps to reach the SDK flow)
   
2. Limited scope: This PR only validates "can the app launch and show a button" vs. the existing infrastructure which tests full payment flows, error handling, and validation logic.

3. Inefficient build strategy: Building apps from scratch on every test run vs. using pre-built artifacts (2-5 min execution). We don't have all the dots connected yet, but this could be an impact point where we actually plan and **connect** them.

Rather than maintaining separate Appium infrastructure per platform, I suggest we extend the existing adyen-checkout-e2e-testing to include React Native:

This aligns with the meta repository vision where every effort scales to all mobile platforms automatically.

---

Note: This is not a blocker for merging - just an invitation to think bigger picture about "SDK development" vs "platform-specific development".

File: .github/workflows/blank_test.yml (hunk 8 above)
Comment (descorp) [reply .github/workflows/blank_test.yml]: Correct. On Ubuntu it happens pretty fast, but on Mac it take some significant time.

**Comment by Robert-SD** [comment]
@descorp @atmamont I really like the addition and will put the PR into perspective. Here are my thoughts:
1. Duplicated Infrastructure

- Clean Build Test: Adding a test for creating a clean empty app project and including the SDK is valuable for verifying that the PR does not break the compilation of the SDK. This is similar to the steps in the Flutter SDK:[ Android workflow](https://github.com/Adyen/adyen-flutter/blob/main/.github/workflows/android_clean_build.yml) and[ iOS workflow](https://github.com/Adyen/adyen-flutter/blob/main/.github/workflows/ios_clean_build.yml).
- Question on Appium: I'm not sure if Appium is truly necessary for this, or if simply building and launching the app without interacting with buttons would suffice.
If Appium is added, we should follow our global Kotlin based approach. We can work together to reuse our centralized element mapping, for which the groundwork has already been prepared with React Native in mind.

2. Limited Scope & E2E Testing
- The full payment flow will be tested through the End-to-End (e2e) project, which uses shared tests across all platforms. This means there is no need to create custom React Native tests besides the steps required to reach the native SDK.
- Netlify Layer: This setup requires the Netlify layer, similar to the MyStore demo apps.
I've added a Netlify network layer to the Flutter example app instead of maintaining two separate apps. So we can switch just the credentials and the endpoints. We could consider doing the same for the React Native example app.

3. Build Strategy & Dedicated Tests
- The e2e tests run as nightlies and are not quick. Using them during the day while implementing features might be too time-consuming. A quick clean build check is nice.
- As mentioned in point 1, I like having a dedicated "building/compilation" test inside the SDK project.  Also for renovate PRs. This is where we could test the latest Gradle and Android versions, as well as the latest iOS version.

File: e2e-tests/scripts/build-android.sh [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+echo "== Build Android"
+cd android || exit
+chmod +x ./gradlew || true
+./gradlew installDebug || bash ./gradlew installDebug
+cd ..
+
+bash ./start_metro.sh
+
+echo "== Reverse Metro Port"
+adb reverse tcp:8081 tcp:8081 || true
```
[the remark this answers is not in the document]
Comment (descorp) [reply e2e-tests/scripts/build-android.sh]: For local run this can be useful if device already connected. Will not change

File: scripts/fixture_ios.sh (hunk 10 above)
Comment (descorp) [reply scripts/fixture_ios.sh]: As discussed, will address in other in other PR.

File: scripts/fixture_android.sh (hunk 11 above)
Comment (descorp) [reply scripts/fixture_android.sh]: Not necessary.

A standard GitHub runner is ephemeral: once the job completes (whether it passes or fails), it automatically attempts to kill any orphaned processes and process trees left running at the end of a job execution before the entire Virtual Machine (VM) is destroyed.

Script: 
```bash
yarn start --port 8081 >/dev/null 2>&1 & # The '&' runs it in background
```

File: e2e-tests/scripts/build-android.sh [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+if [ "$platform" == "Expo" ]; then
+  echo "== Prebuild Expo Android"
+  npx expo prebuild -p android --clean
+fi
+
+echo "== Build Android"
+cd android || exit
+chmod +x ./gradlew || true
+./gradlew installDebug || bash ./gradlew installDebug
```
[the remark this answers is not in the document]
Comment (descorp) [reply e2e-tests/scripts/build-android.sh]: Addressed

File: e2e-tests/scripts/inject_secrets.sh [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  exit 1
+fi
+
+echo "Injecting secrets into $SECRETS_FILE"
+
+# We use '#' as delimiter, so secrets can safely contain '|'.
+# Keys are expected to only contain [a-zA-Z0-9_|], so no extra escaping is needed.
+sed -e "s#__CLIENT_KEY__#${ADYEN_CLIENT_KEY}#g" \
+    -e "s#__PUBLIC_KEY__#${ADYEN_PUBLIC_KEY}#g" \
+    "$SECRETS_FILE" > "$SECRETS_FILE.tmp" && mv "$SECRETS_FILE.tmp" "$SECRETS_FILE"
```
[the remark this answers is not in the document]
Comment (descorp) [reply e2e-tests/scripts/inject_secrets.sh:29]: I do not contain # symbol.

File: e2e-tests/template/App.tsx.template [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+  const { start } = useAdyenCheckout();
+
+  return (
+    <View>
+      <Button
+        testID="dropin-button"
+        title="Open DropIn"
+        onPress={async () => {
+          start('dropin');
+        }}
```
[the remark this answers is not in the document]
Comment (descorp) [reply e2e-tests/template/App.tsx.template:99]: Will fix in next PR.

File: e2e-tests/scripts/start_metro.sh [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+for i in {1..45}; do
+  if nc -z 127.0.0.1 "$PORT"; then
+    echo "== Metro started successfully"
+    exit 0
+  fi
+  sleep 1
+done
+
+echo "Error: Metro failed to start within 60 seconds"
```
[the remark this answers is not in the document]
Comment (descorp) [reply e2e-tests/scripts/start_metro.sh:19]: Addressed in [c8403c1](https://github.com/Adyen/adyen-react-native/pull/880/commits/c8403c13d5f1b964f27eb2659586760e0fc9a9d1)

File: e2e-tests/scripts/build-android.sh [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+if [ "$platform" = "expo" ]; then
+  echo "::group::Prebuild Expo Android [$(date '+%H:%M:%S')]"
+  npx expo prebuild -p android --clean
+  echo "::endgroup::"
+fi
+
+echo "::group::Build Android [$(date '+%H:%M:%S')]"
+cd android || exit
+chmod +x ./gradlew
+./gradlew installDebug || bash ./gradlew installDebug
```
[the remark this answers is not in the document]
Comment (descorp) [reply e2e-tests/scripts/build-android.sh]: Addressed in [c8403c1](https://github.com/Adyen/adyen-react-native/pull/880/commits/c8403c13d5f1b964f27eb2659586760e0fc9a9d1)

File: package.json [hunk 18]
```diff
@@ -113,10 +106,11 @@
...
     "preset": "react-native",
+    "testPathIgnorePatterns": [
+      "e2e-tests"
+    ],
     "modulePathIgnorePatterns": [
-      "_mock_",
-      "<rootDir>/example/node_modules",
-      "<rootDir>/lib/"
+      "_mock_"
     ]
```
[the remark this answers is not in the document]
Comment (descorp) [reply package.json:116]: Addressed in [c8403c1](https://github.com/Adyen/adyen-react-native/pull/880/commits/c8403c13d5f1b964f27eb2659586760e0fc9a9d1)

**Review by Robert-SD — APPROVED** [review]
